### PR TITLE
Fixes segfault when opening headlines #828 (reported by chronoszc)

### PR DESCRIPTION
**Pull request by lwindolf**
This should fix the comment feed update segfault by converting the updateRequest structure to a GObject that is properly references and unreferenced, so that there can be concurrent "consumers" of a request object.

### CODE REVIEW

**Comment by Leiaz** [comment]
It doesn't actually fix the segfault for me.
`state` is NULL in `update_state_get_lastmodified`, when updating a comment thread with 0 comments and receiving a 304.

Downloading a comment thread with comments works : 
```
NET    : downloading https://delphi.cz/syndication.axd?post=7a7a8b92-7a13-4cf4-acc2-75677955749a
> GET /syndication.axd?post=7a7a8b92-7a13-4cf4-acc2-75677955749a HTTP/1.1
> Soup-Debug-Timestamp: 1590237816
> Soup-Debug: SoupSession 1 (0x555555d1d100), SoupMessage 1 (0x555556ad78c0), SoupSocket 1 (0x555555cbf740)
> Host: delphi.cz
> Accept: application/atom+xml,application/xml;q=0.9,text/xml;q=0.8,*/*;q=0.7
> DNT: 1
> Accept-Encoding: gzip, deflate
> User-Agent: Liferea/1.13.1 (Linux; https://lzone.de/liferea/) AppleWebKit (KHTML, like Gecko)
> Connection: Keep-Alive
  
< HTTP/1.1 200 OK
< Soup-Debug-Timestamp: 1590237816
< Soup-Debug: SoupMessage 1 (0x555556ad78c0)
< Cache-Control: public
< Content-Type: application/rss+xml
< Content-Encoding: deflate
< Last-Modified: Fri, 22 Nov 2019 22:58:29 GMT
< ETag: "637100639090000000"
< Server: Microsoft-IIS/8.5
< Content-Disposition: inline; filename=rss.xml
< X-Powered-By: ASP.NET
< X-Powered-By: ARR/3.0
< Date: Sat, 23 May 2020 12:43:34 GMT
< Content-Length: 1448
  
NET    : download status code: 200
NET    : source after download: >>>https://delphi.cz/syndication.axd?post=7a7a8b92-7a13-4cf4-acc2-75677955749a<<<
NET    : 6466 bytes downloaded
```

Downloading a comment threads with no comments results in a segfault : 
```
NET    : downloading https://delphi.cz/syndication.axd?post=b6a6a839-0272-407d-9be9-3ece0c3ce087
> GET /syndication.axd?post=b6a6a839-0272-407d-9be9-3ece0c3ce087 HTTP/1.1
> Soup-Debug-Timestamp: 1590237818
> Soup-Debug: SoupSession 1 (0x555555d1d100), SoupMessage 2 (0x555556ad7aa0), SoupSocket 1 (0x555555cbf740)
> Host: delphi.cz
> Accept: application/atom+xml,application/xml;q=0.9,text/xml;q=0.8,*/*;q=0.7
> DNT: 1
> Accept-Encoding: gzip, deflate
> User-Agent: Liferea/1.13.1 (Linux; https://lzone.de/liferea/) AppleWebKit (KHTML, like Gecko)
> Connection: Keep-Alive
  
< HTTP/1.1 304 Not Modified
< Soup-Debug-Timestamp: 1590237818
< Soup-Debug: SoupMessage 2 (0x555556ad7aa0)
< Cache-Control: public
< Content-Type: application/rss+xml
< Last-Modified: Mon, 01 Jan 0001 00:00:00 GMT
< ETag: "0"
< Server: Microsoft-IIS/8.5
< Content-Disposition: inline; filename=rss.xml
< X-Powered-By: ASP.NET
< X-Powered-By: ARR/3.0
< Date: Sat, 23 May 2020 12:43:37 GMT
  
NET    : download status code: 304
NET    : source after download: >>>https://delphi.cz/syndication.axd?post=b6a6a839-0272-407d-9be9-3ece0c3ce087<<<
NET    : 0 bytes downloaded

Thread 1 "liferea" received signal SIGSEGV, Segmentation fault.
update_state_get_lastmodified (state=0x0) at ../update.c:72
72		return state->lastModified;
(gdb) bt
#0  update_state_get_lastmodified (state=0x0) at ../update.c:72
#1  0x00005555555858f5 in network_process_callback
    (session=<optimized out>, msg=0x555556ad7aa0, user_data=0x555556e2efc0)
    at ../net.c:102
#2  0x00007ffff499dd8d in  () at /lib/x86_64-linux-gnu/libsoup-2.4.so.1
#3  0x00007ffff499e7a8 in  () at /lib/x86_64-linux-gnu/libsoup-2.4.so.1
#4  0x00007ffff499e846 in  () at /lib/x86_64-linux-gnu/libsoup-2.4.so.1
#5  0x00007ffff3cae4de in g_main_context_dispatch ()
    at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#6  0x00007ffff3cae890 in  () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#7  0x00007ffff3cae91f in g_main_context_iteration ()
    at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#8  0x00007ffff3efa05d in g_application_run ()
    at /lib/x86_64-linux-gnu/libgio-2.0.so.0
#9  0x0000555555584949 in liferea_application_new (argc=2, argv=0x7fffffffe598)
    at ../liferea_application.c:351
#10 0x00007ffff395be0b in __libc_start_main (main=
    0x555555574320 <main>, argc=2, argv=0x7fffffffe598, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffe588)
    at ../csu/libc-start.c:308
#11 0x00005555555743da in _start () at main.c:54
(gdb) f 1
#1  0x00005555555858f5 in network_process_callback (session=<optimized out>, 
    msg=0x555556ad7aa0, user_data=0x555556e2efc0) at ../net.c:102
102			 job->result->updateState->lastModified = update_state_get_lastmodified (job->request->updateState);
```

File: src/update.c [hunk 1]
```diff
@@ -147,27 +147,39 @@ update_state_free (updateStatePtr updateState)
...
+	UpdateRequest *request = UPDATE_REQUEST (obj);
 
 	update_state_free (request->updateState);
 	update_options_free (request->options);
 
 	g_free (request->postdata);
 	g_free (request->source);
 	g_free (request->filtercmd);
-	g_free (request);
+}
```
Comment (Leiaz) [inline src/update.c:165]: I think the parent class's finalized should be called here. Although I'm not sure if GObject's finalize does anything ..

File: src/update.c [hunk 2]
```diff
@@ -247,7 +261,7 @@ update_job_new (gpointer owner,
 
 	job = g_new0 (struct updateJob, 1);
 	job->owner = owner;
-	job->request = request;
+	job->request = UPDATE_REQUEST (g_object_ref (request));
```
Comment (Leiaz) [inline src/update.c:264]: Previously, it looked like the `updateJob` took ownership of the request. Here the job takes a reference on the request, but there is already a reference and the caller doen't unref it, so it looks like the UpdateRequest will be leaked, as the reference count will still be one after `update_job_free`.

File: src/update.h [hunk 3]
```diff
@@ -117,7 +122,7 @@ typedef struct updateResult {
 
 /** structure describing an HTTP update job */
 typedef struct updateJob {
-	updateRequestPtr	request;
+	UpdateRequest		*request;
 	updateResultPtr		result;
```
Comment (Leiaz) [inline src/update.h:126]: You have replaced `updateRequestPtr` in some places but not others. But maybe it is intentional ?

File: src/update.c (hunk 1 above)
Comment (lwindolf) [reply src/update.c:167]: Added it just to be sure!

File: src/update.c (hunk 2 above)
Comment (lwindolf) [reply src/update.c]: Also fixed. Thanks for catching this!

**Comment by Leiaz** [comment]
It still doesn't solve the segfault for me. So the changelog is wrong. But other than that I see no issue.

**Comment by lwindolf** [comment]
Damn... have to fix this! Thanks for your patience with testing!

**Comment by Leiaz** [comment]
The segfault is caused by the `request->updateState` being NULL.
It looks like, unlike subscriptions, with comment feeds the updateState is not copied to the request. So a basic fix would be this :
```diff
diff --git a/src/comments.c b/src/comments.c
index 86f28803..dc473ac6 100644
--- a/src/comments.c
+++ b/src/comments.c
@@ -225,6 +225,7 @@ comments_refresh (itemPtr item)
 		request = update_request_new ();
 		request->options = g_new0 (struct updateOptions, 1);	// FIXME: use copy of parent subscription options
 		request->source = g_strdup (url);
+		request->updateState = update_state_copy (commentFeed->updateState);
 		commentFeed->updateJob = update_execute_request (commentFeed, request, comments_process_update_result, commentFeed, FEED_REQ_PRIORITY_HIGH);
 
 		/* Item view refresh to change link from "Update" to "Updating..." */
```
But maybe some of this initialization of the UpdateRequest structure should happen in `update_request_new`, passing the required data, like the source, the options ... It would make it clear what must be set or what could be null.

**Comment by lwindolf** [comment]
@Leiaz Thank you for the hint. I added it and refactored update_request_new() to ensure consistent setting of relevant members.

File: src/comments.c [hunk 4]
```diff
@@ -175,60 +175,63 @@ comments_process_update_result (const struct updateResult * const result, gpoint
...
 				commentFeeds = g_hash_table_new (g_str_hash, g_str_equal);
 			g_hash_table_insert (commentFeeds, commentFeed->id, commentFeed);
 		}
 
-		request = update_request_new ();
-		request->options = g_new0 (struct updateOptions, 1);	// FIXME: use copy of parent subscription options
-		request->source = g_strdup (url);
+		request = update_request_new (
+			url,
+			commentFeed->updateState,
```
Comment (Leiaz) [inline src/comments.c:226]: You removed `update_state_new` above, so here `commentFeed->updateState` could be uninitialized.

File: src/comments.c [hunk 5]
```diff
@@ -175,60 +175,63 @@ comments_process_update_result (const struct updateResult * const result, gpoint
...
+			NULL	// FIXME: use copy of parent subscription options
+		);
+
+		commentFeed->updateState = request->updateState;
```
Comment (Leiaz) [inline src/comments.c:230]: Here you set `commentFeed->updateState` to the request's update state that will be destroyed after the job terminates. So it still segfaults when `update_request_new` tries to copy something that isn't there.
`commentFeed` should have its own copy of update state to keep the state between update jobs. I see you made `update_state_new` static but in subscription.c there is a `g_new0 (struct updateState, 1)`.

File: src/comments.c (hunk 4 above)
Comment (lwindolf) [reply src/comments.c:228]: This should be ok, as it is NULL in this case and update_request_new() creates a new one.

Together with an additional fix in the result processing callback, where I now explicitely copy the result update state, this should be fine.

File: src/comments.c (hunk 5 above)
Comment (lwindolf) [reply src/comments.c]: You are right. I dropped this line. It doesn't make any sense at all. The actual explicit copying now happens in the result callback.

I also fixed the g_new0() in subscription.c

**Review by Leiaz — APPROVED** [review]
Looks good.

**Pull request merged**